dash-renderer/src/reducers/callbacks.ts

```diff
@@ -0,0 +1,147 @@
+import {
+    concat,
+    difference,
+    reduce
+} from 'ramda';
+
+import {
+    ICallback,
+    IExecutedCallback,
+    IExecutingCallback,
+    IStoredCallback
+} from '../types/callbacks';
+
+export enum CallbackActionType {
+    AddApplied = 'Callbacks.AddApplied',
+    AddExecuted = 'Callbacks.AddExecuted',
+    AddExecuting = 'Callbacks.AddExecuting',
+    AddPrioritized = 'Callbacks.AddPrioritized',
+    AddRequested = 'Callbacks.AddRequested',
+    AddStored = 'Callbacks.AddStored',
+    AddWatched = 'Callbacks.AddWatched',
+    RemoveApplied = 'Callbacks.RemoveApplied',
+    RemoveExecuted = 'Callbacks.RemoveExecuted',
+    RemoveExecuting = 'Callbacks.RemoveExecuting',
+    RemovePrioritized = 'Callbacks.ReomvePrioritized',
+    RemoveRequested = 'Callbacks.RemoveRequested',
+    RemoveStored = 'Callbacks.RemoveStored',
+    RemoveWatched = 'Callbacks.RemoveWatched'
+}
+
+export enum CallbackAggregateActionType {
+    AddCompleted = 'Callbacks.Completed',
+    Aggregate = 'Callbacks.Aggregate'
+}
+
+export interface IAggregateAction {
+    type: CallbackAggregateActionType.Aggregate,
+    payload: (ICallbackAction | ICompletedAction | null)[]
+}
+
+export interface ICallbackAction {
+    type: CallbackActionType;
+    payload: ICallback[];
+}
+
+export interface ICompletedAction {
+    type: CallbackAggregateActionType.AddCompleted,
+    payload: number
+}
+
+type CallbackAction =
+    IAggregateAction |
+    ICallbackAction |
+    ICompletedAction;
+
+export interface ICallbacksState {
+    requested: ICallback[];
+    prioritized: ICallback[];
+    executing: IExecutingCallback[];
+    watched: IExecutingCallback[];
+    executed: IExecutedCallback[];
+    stored: IStoredCallback[];
+    completed: number;
+}
```

Comment:
I gather this is the order that callbacks step through the process? I think I understand some of it:
- requested: something has caused this callback to need to fire
- prioritized: we've determined that this callback isn't waiting on anything other callback to complete
- executing: we have initiated the request for this callback

Is that correct? What are the other three (watched, executed, stored) for?

Reply:
`stored` holds on to `executed` callbacks after the response has been processed. Once there are no more callbacks associated with the execution group (layout and initialization callbacks) the stored callbacks for an execution group are removed.

`executed` callbacks have been processed by the server but their response has not been processed into the layout yet.

`watched` callbacks are probably redundant with `executing` callbacks - I'll see if I can 🔪  that step out entirely.

`executing`, `prioritized` and `requested` callbacks are as you describe above.